### PR TITLE
export groups (#487)

### PR DESCRIPTION
**Pull request by retorquere**

### CODE REVIEW

**Comment by dstillman** [comment]
This would need to be based on libraryID, not specific to groups. (libraryID is null for My Library in 4.0.)

**Comment by retorquere** [comment]
Right, but this would only show up for groups, not for libraries. Library export is handled as it was before.

**Comment by dstillman** [comment]
There will be other types of libraries in 5.0.

**Comment by dstillman** [comment]
And there's also no reason the Export Library on My Library shouldn't run through the same code.

**Comment by retorquere** [comment]
I just figured collections also get a "Export Collection" menu item, and the UI calls it a group, not a library. There's also that group export really does do something else -- library export is inferred by the Zotero code from the absence of limiting filters (groups/collections/items) and the collection-gathering is different for both (but only from a technical POV). Your call though.

File: chrome/content/zotero/xpcom/translation/translate_item.js [hunk 1]
```diff
@@ -786,6 +786,19 @@ Zotero.Translate.ItemGetter.prototype = {
 		this.numItems = this._itemsLeft.length;
 	},
 	
+	"setGroup":function(group, getChildCollections) {
+		this._itemsLeft = Zotero.DB.columnQuery('select itemID from items where libraryID in (select libraryID from groups where groupID = ?) and not itemID in (select itemID from deletedItems)', [group.id]);
```
Comment (dstillman) [inline chrome/content/zotero/xpcom/translation/translate_item.js]: This will all need to be redone to not be group-specific, but there shouldn't be any SQL code in here. It should call Zotero.Items.getAll(), as setAll() does below, but with a libraryID for the second parameter. (I'm fixing the messy argument order in 5.0.)

**Comment by dstillman** [comment]
Well, the UI calls them Group Libraries. And again, there will be different library types in 5.0, so this should work on those too without any new code.

**Comment by retorquere** [comment]
OK, I'll get on that. I'll need some help squashing commits when I'm done -- never did get the hang of that.

**Comment by dstillman** [comment]
No problem.

Note that the existing interfaces likely predate the existence of other libraries, so they can be changed. E.g., you can rename setAll() to setLibrary() and include a libraryID in the call. Zotero.getCollections() also can take a libraryID. (The API here is messy and has generally been cleaned up (e.g., Zotero.Collections.getCollectionsInLibrary()) on the api_syncing branch, which will become 5.0, but translation is currently broken there, which is why I'll allow this on 4.0.)

**Comment by retorquere** [comment]
OK, I have it working, how do I squash the commit?

**Comment by aurimasv** [comment]
Don't worry about squashing for now, just push the additional commits. It makes it easier for us to review. You can squash once everything looks right.

**Comment by retorquere** [comment]
done

**Comment by dstillman** [comment]
This doesn't seem to work. Looks like it's exporting all items, not just those from the selected group.

File: chrome/content/zotero/fileInterface.js [hunk 2]
```diff
@@ -129,7 +131,8 @@ var Zotero_File_Interface = new function() {
 	 */
 	function exportFile() {
```
Comment (aurimasv) [inline chrome/content/zotero/fileInterface.js]: nit: we should rename this to something like `exportLibrary`, since this name is confusing and inconsistent with functions below. As you're renaming, you can also just use the `this.exportLibrary = function() {...` convention, since that is what Zotero is slowly transitioning to (as opposed to declaring the functions at the top of the object as is done here)

File: chrome/content/zotero/fileInterface.js [hunk 3]
```diff
@@ -76,6 +76,8 @@ Zotero_File_Exporter.prototype.save = function() {
 			translation.setCollection(this.collection);
 		} else if(this.items) {
 			translation.setItems(this.items);
+		} else if(this.libraryID) {
+			translation.setLibraryID(this.LibraryID);
```
Comment (aurimasv) [inline chrome/content/zotero/fileInterface.js]: lower case

File: chrome/content/zotero/fileInterface.js (hunk 2 above)
Comment (dstillman) [inline chrome/content/zotero/fileInterface.js]: That doesn't need to happen for this patch, though. 4.0 isn't the place to make unnecessary changes, since it will just result in lots of conflicts.

**Comment by retorquere** [comment]
Hey, you're right. I hadn't noticed because my test group lib has the same items as the main one. I'll investigate.

File: chrome/content/zotero/fileInterface.js [hunk 4]
```diff
@@ -129,7 +131,8 @@ var Zotero_File_Interface = new function() {
 	 */
 	function exportFile() {
 		var exporter = new Zotero_File_Exporter();
-		exporter.name = Zotero.getString("pane.collections.library");
+		exporter.libraryID = ZoteroPane_Local.getSelectedLibraryID();
```
Comment (aurimasv) [inline chrome/content/zotero/fileInterface.js]: This could return false if something other than a library is selected (e.g. "Group Libraries" label). I don't think we should be able to get to this point in those circumstances, so you should throw if that's the case.

File: chrome/content/zotero/fileInterface.js (hunk 3 above)
Comment (retorquere) [inline chrome/content/zotero/fileInterface.js]: What would you prefer to see lowercased? The "ID"?

Comment (aurimasv) [inline chrome/content/zotero/fileInterface.js]: `libraryID` (this is why this is not working btw)

Comment (retorquere) [inline chrome/content/zotero/fileInterface.js]: Ah, I see now, thanks!

**Comment by retorquere** [comment]
Looks like _libraryID is in use in translate.js and gets overwritten. What varname can I use instead?

**Comment by aurimasv** [comment]
I'm guessing it's happening here: https://github.com/zotero/zotero/blob/4.0/chrome/content/zotero/xpcom/translation/translate.js#L1173 The variable name seems appropriate, I just think that you need to make sure to pass the library ID into `translate` at some point. Don't have time right now to give you more guidance though.

File: chrome/content/zotero/xpcom/translation/translate_item.js [hunk 5]
```diff
@@ -1034,4 +1034,4 @@ Zotero.Translate.ItemGetter.prototype = {
 		return obj;
 	}
 }
-Zotero.Translate.ItemGetter.prototype.__defineGetter__("numItemsRemaining", function() { return this._itemsLeft.length });
\ No newline at end of file
+Zotero.Translate.ItemGetter.prototype.__defineGetter__("numItemsRemaining", function() { return this._itemsLeft.length });
```
Comment (aurimasv) [inline chrome/content/zotero/xpcom/translation/translate_item.js]: Please revert this to not pollute git blame

**Comment by retorquere** [comment]
I've gone another route in https://github.com/zotero/zotero/commit/9a0820cca2f300c02da565d6d437460301a3f952

**Comment by retorquere** [comment]
This also fixes the problem that the main library gets exported instead of the group library. It didn't happen before as I was passing around a group in a differently-named variable.

File: chrome/content/zotero/fileInterface.js (hunk 4 above)
Comment (retorquere) [inline chrome/content/zotero/fileInterface.js]: done

Comment (retorquere) [inline chrome/content/zotero/fileInterface.js]: I've also chomped the trailing newlines

File: chrome/content/zotero/xpcom/translation/translate_item.js (hunk 5 above)
Comment (retorquere) [inline chrome/content/zotero/xpcom/translation/translate_item.js]: done

File: chrome/content/zotero/fileInterface.js [hunk 6]
```diff
@@ -129,7 +131,9 @@ var Zotero_File_Interface = new function() {
...
+		if exporter.libraryID === false throw('No library selected');
```
Comment (aurimasv) [inline chrome/content/zotero/fileInterface.js]: Are you testing this? I'm fairly certain that's a syntax error. Can't test it right now, so in the off chance that it's not, please use parentheses for conditionals.

Comment (aurimasv) [inline chrome/content/zotero/fileInterface.js]: Also, `throw new Error("...")` so it generates a stack trace

File: chrome/content/zotero/xpcom/translation/translate.js [hunk 7]
```diff
@@ -2182,15 +2188,21 @@ Zotero.Translate.Export.prototype._prepareTranslation = function() {
...
 		getCollections = configOptions.getCollections || false;
-	if(this._collection) {
-		this._itemGetter.setCollection(this._collection, getCollections);
-		delete this._collection;
-	} else if(this._items) {
-		this._itemGetter.setItems(this._items);
-		delete this._items;
-	} else {
-		this._itemGetter.setAll(getCollections);
+	switch ((this._export || {}).type) {
```
Comment (aurimasv) [inline chrome/content/zotero/xpcom/translation/translate.js]: It should be an error to call this without setting `this._export` first (may have to fix that in current code). Likewise, it should be an error to call this with unknown `this._export.type`, so the `default` should throw.

File: chrome/content/zotero/xpcom/translation/translate.js [hunk 8]
```diff
@@ -2182,15 +2188,21 @@ Zotero.Translate.Export.prototype._prepareTranslation = function() {
...
+			this._itemGetter.setCollection(this._export.collection, getCollections);
+			break;
+		case 'items':
+			this._itemGetter.setItems(this._export.items);
+			break;
+		case 'library':
+			this._itemGetter.setAll(this._export.id, getCollections);
+			break;
+		default:
+			this._itemGetter.setAll();
```
Comment (aurimasv) [inline chrome/content/zotero/xpcom/translation/translate.js]: You forgot the `getCollections` argument.

File: chrome/content/zotero/xpcom/translation/translate_item.js [hunk 9]
```diff
@@ -786,11 +786,11 @@ Zotero.Translate.ItemGetter.prototype = {
...
 	},
 	
-	"setAll":function(getChildCollections) {
-		this._itemsLeft = Zotero.Items.getAll(true);
+	"setAll":function(libraryID, getChildCollections) {
+		this._itemsLeft = Zotero.Items.getAll(true, libraryID);
 		
 		if(getChildCollections) {
-			this._collectionsLeft = Zotero.getCollections();
+			this._collectionsLeft = Zotero.getCollections(null, false, libraryID);
```
Comment (aurimasv) [inline chrome/content/zotero/xpcom/translation/translate_item.js]: The second parameter here should be `true` (i.e. recursively) as is the case for `setCollection` above. Looks like this was a bug.

File: chrome/content/zotero/fileInterface.js (hunk 2 above)
Comment (aurimasv) [inline chrome/content/zotero/fileInterface.js]: Do you mean changing the function name or the second part or both?

File: chrome/content/zotero/xpcom/translation/translate_item.js (hunk 9 above)
Comment (retorquere) [inline chrome/content/zotero/xpcom/translation/translate_item.js]: I had been wondering about that.

File: chrome/content/zotero/fileInterface.js (hunk 6 above)
Comment (retorquere) [inline chrome/content/zotero/fileInterface.js]: Dammit, the extension pointed to a different repo checkout. I've nuked all but one version.

File: chrome/content/zotero/xpcom/translation/translate.js (hunk 7 above)
Comment (retorquere) [inline chrome/content/zotero/xpcom/translation/translate.js]: Alright, I was just trying to be cautious here.

File: chrome/content/zotero/xpcom/translation/translate.js (hunk 8 above)
Comment (retorquere) [inline chrome/content/zotero/xpcom/translation/translate.js]: But that could just be removed as the default is going to throw anyway

**Comment by retorquere** [comment]
OK, latest comments have been addressed. Is there any way in which I can help building up the testing infrastructure? I have really gotten used to automated unit tests.

File: chrome/content/zotero/fileInterface.js (hunk 2 above)
Comment (dstillman) [inline chrome/content/zotero/fileInterface.js]: Both. Any remaining patches on 4.0 need to stay as small as possible.

Comment (retorquere) [inline chrome/content/zotero/fileInterface.js]: So no action on my part?

Comment (dstillman) [inline chrome/content/zotero/fileInterface.js]: That's right.

**Comment by retorquere** [comment]
Oh and if someone could walk me through the squash, that would be very much appreciated.

File: chrome/content/zotero/xpcom/translation/translate.js [hunk 10]
```diff
@@ -2182,15 +2188,22 @@ Zotero.Translate.Export.prototype._prepareTranslation = function() {
...
 		getCollections = configOptions.getCollections || false;
-	if(this._collection) {
-		this._itemGetter.setCollection(this._collection, getCollections);
-		delete this._collection;
-	} else if(this._items) {
-		this._itemGetter.setItems(this._items);
-		delete this._items;
-	} else {
-		this._itemGetter.setAll(getCollections);
+	if (!this._export) { throw new Error('No export set up'); }
```
Comment (aurimasv) [inline chrome/content/zotero/xpcom/translation/translate.js]: you don't need explicit `throw` here. It will throw implicitly from the `switch` if `this._export` is not set.

Comment (retorquere) [inline chrome/content/zotero/xpcom/translation/translate.js]: Only if this._export is set, right? Otherwise it will throw a "this._export is null" error. Or is that the desired effect? Easily removed.

**Comment by dstillman** [comment]
> Is there any way in which I can help building up the testing infrastructure? I have really gotten used to automated unit tests.

It's not worth writing tests for most things on the 4.0 branch because most of the API is changing with [promisification](https://github.com/zotero/zotero/issues/518). But the api_syncing branch, which will become 5.0, has [~100 tests](https://github.com/zotero/zotero/tree/api_syncing/test/tests) and counting. That branch isn't quite ready for public development (which is why it's not on master yet), but it will be very soon, after which we'll require tests for most things.

File: chrome/content/zotero/fileInterface.js [hunk 11]
```diff
@@ -76,6 +76,8 @@ Zotero_File_Exporter.prototype.save = function() {
 			translation.setCollection(this.collection);
 		} else if(this.items) {
 			translation.setItems(this.items);
+		} else {
+			translation.setLibraryID(this.libraryID);
```
Comment (aurimasv) [inline chrome/content/zotero/fileInterface.js]: Let's make sure that `this.libraryID` is set and throw if none of the above are set. Currently, all code sets one of these, but it will probably make debugging a lot easier if we forget at some point. (throw something like "collection, items, or libraryID must be set before calling Zotero_File_Exporter#save()")

**Comment by aurimasv** [comment]
That's all I have right now. If @dstillman has no other comments, you can squash after making these changes. http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html may be helpful. We can be more specific if you need more guidance.

File: chrome/content/zotero/fileInterface.js (hunk 11 above)
Comment (retorquere) [inline chrome/content/zotero/fileInterface.js]: How would we distinguish between main library export and libraryID not being set?

File: chrome/content/zotero/xpcom/translation/translate.js (hunk 10 above)
Comment (aurimasv) [inline chrome/content/zotero/xpcom/translation/translate.js]: JavaScript engine throwing "cannot read property '_export' of null" is good enough here, since we get a stack anyway.

File: chrome/content/zotero/fileInterface.js (hunk 11 above)
Comment (retorquere) [inline chrome/content/zotero/fileInterface.js]: done.

Comment (retorquere) [inline chrome/content/zotero/fileInterface.js]: What's the ETA on 5.0 BTW? Should I be scrambling to rewrite BBT?

**Comment by retorquere** [comment]
What should I make of the warning at the top of that gitready page: "Only do this on commits that haven’t been pushed an external repository"?

**Comment by aurimasv** [comment]
It's only relevant if someone else forks the branch in question and does further work on it. That could really mess them up. In case of GitHub topic branches where you are the only committer, that's not relevant.

**Comment by retorquere** [comment]
Can someone verify the squash went OK?

**Comment by aurimasv** [comment]
No, that didn't work. What command exactly did you run?

**Comment by dstillman** [comment]
Before trying something else, I would recommend using git reflog, finding the commit you were at before that rebase (i.e., the last commit you made on this branch), and doing a git reset --hard to that commit.

**Comment by aurimasv** [comment]
I think your local 4.0 branch is out of date. Make sure you either rebase against upstream/4.0 or you pull upstream changes into your local 4.0 branch

**Comment by dstillman** [comment]
(Or just rebase against the first commit in the topic branch that isn't yours, such that you only see your own 8 commits in the interactive list (after doing a hard reset to undo the above). Rebasing against upstream again isn't usually necessary if you checked out the topic branch from the upstream branch recently, and GitHub will tell you on the pull request when it can't do a clean merge.)

**Comment by retorquere** [comment]
Ugh I hate this part of git. I did `git rebase -i HEAD~8`, and squashed all but the top one. I'm sorry but I don't understand the terminology starting from "before trying anything else" on. I have a patch file that has just my changes, at this point it sounds easier to just get a fresh checkout, apply the patch, and push that.

**Comment by dstillman** [comment]
I understand this can be tricky, but we really prefer keeping these things on the original pull requests.

**Comment by dstillman** [comment]
Make a backup of your repo. Then run `git reflog`, and look down the list until you see the 8 commits you had before, which should be before some rebase junk. Find the hash of the most recent commit. Then do a `git reset --hard <that commit>`. Then you should be back where you were before, and you can do a push -f from there. Then let's pause and reevaluate.

**Comment by dstillman** [comment]
(The commit to hard-reset to might be 0b36e99963834025bcc47b97b3adf3b11b519b87, based on the fetch I did right before you did this.)

**Comment by dstillman** [comment]
[Disregard my deleted message. I think I fetched after you rebased.]

**Comment by retorquere** [comment]
OK, I'm there. Should we move this to gitter?

**Comment by dstillman** [comment]
OK, then force-push to get those 8 commits back on this branch.

**Comment by retorquere** [comment]
OK, that should be there now.

**Comment by aurimasv** [comment]
(if you want live chat, you can join zotero's IRC channel irc://irc.freenode.net/zotero-dev )

**Comment by retorquere** [comment]
I don't mind either way. The pull request now says "10 files changed", that cannot be right.

**Comment by dstillman** [comment]
Join zotero-dev and we'll walk you through this.

**Comment by dstillman** [comment]
Thanks!
